File: String/Cpp/2-Valid_Anagram.cpp
// VALID ANAGRAM

#include<bits/stdc++.h>
using namespace std;

bool validAnagram(string input1, string input2){
    if(input1.length() != input2.length()){
        return false;
    }
    
    /* // BRUTE FORCE APPROCH USING SORTING (TIME-O(nlogn), SPACE-O(1))
    sort(input1.begin(), input1.end());
    sort(input2.begin(), input2.end());
    if(input1 == input2) return true;
    else return false;
    */
    
    /* // BETTER APPROCH 1 - USING HASHMAP OR DICTIONARY (TIME-O(n), SPACE-O(1))
    unordered_map<char, int> mpp;
    for(char c : input1){
        mpp[c]++;
    }
    for(char c : input2){
        mpp[c]--;
    }
    for(auto it : mpp){
        if(it.second != 0) return false;
    }
    return true;
    */
    
    // BETTER APPROCH 2 - USING FREQUENCY ARRAY (TIME-O(n), SPACE-O(1))
    int arr[26] = {0};
    for(char c : input1){
        arr[c - 'a']++;
    }
    for(char c : input2){
        arr[c - 'a']--;
    }
    for(char indx : arr){
        if(indx != 0) return false;
    }
    return true;
}

int main(){
    
    // taking input string from user
    string input1, input2;
    cout << "Enter Input String 1 : ";
    cin >> input1;
    cout << "Enter Input String 2 : ";
    cin >> input2;
    
    // calling function
    bool ans = validAnagram(input1, input2);
    
    // output
    if(ans) cout << "true" << endl;
    else cout << "false" << endl;
    
}

File: String/Python/2-Valid_Anagram.py
# VALID ANAGRAM

def validAnagram(input1 : str, input2 : str) -> bool:
    if( len(input1) != len(input2) ):
        return False
        
    ''' # BRUTE FORCE USING SORTING (TIME-O(nlogn), SPACE-O(1))
    input1 = sorted(input1)
    input2 = sorted(input2)
    if( input1 == input2 ):
        return True
    else:
        return False
    '''
    
    '''' # BETTER APPROCH 1 USING HASHMAP OR DICTIONARY (TIME-O(n), SPACE-O(1))
    #or dic = defaultdict(int) which will Initialize a defaultdict with int, which defaults to 0 
    dic = {}     
    for ch in input1:
        if(ch in dic):
            dic[ch] += 1
        else:
            dic[ch] = 1
    for ch in input2:
        if(ch in dic):
            dic[ch] -= 1
        else:
            dic[ch] = 1
    for ch in dic:
        if( dic[ch] != 0 ):
            return False
    return True
    '''
    
    # BETTER APPROCH 2 - USING FREQUENCY ARRAY (TIME-O(n), SPACE-O(1))
    arr = [0] * 26
    for ch in input1:
        arr[ord(ch) - ord('a')] += 1
    for ch in input2:
        arr[ord(ch) - ord('a')] -= 1
    for i in arr:
        if i != 0:
            return False
    return True
    
        
        
input1 = input("Enter Input String 1 : ")
input2 = input("Enter Input String 2 : ")

print(validAnagram(input1, input2))
    
